### PR TITLE
feature/khweb-19-add infinite scroll to LaunchList

### PR DESCRIPTION
**Pull request by vovkis13**
[feature/infinity-loader](https://jira.softserve.academy/browse/KHWEB-19)

 - add infinity loader to LaunchList component on Homepage
 - add smooth scroll to launch list title on LaunchFilter link click

### CODE REVIEW

File: src/components/Home/actions/getLaunches.ts [hunk 1]
```diff
@@ -5,17 +5,20 @@ import {
...
+  (filterValue: Launches, page: number = fetchDefaults.FIRST_PAGE) =>
   async (dispatch: Dispatch<LaunchAction>): Promise<void> => {
-    dispatch(launchRequestAction(filterValue));
+    const isScrollLoading: boolean = page !== 1;
+    dispatch(launchRequestAction(filterValue, isScrollLoading));
     try {
-      const response: LaunchResponse = await fetchLaunches(filterValue);
-      dispatch(launchSuccessAction(filterValue, response.data.results));
+      const response: LaunchResponse = await fetchLaunches(filterValue, page);
+      const { results, count }: LaunchResponse['data'] = response.data;
```
Comment (kerrimov) [inline src/components/Home/actions/getLaunches.ts:20]: Providing type like this not a good idea

File: src/components/LaunchDetailsPage/selectors/selectors.ts [hunk 2]
```diff
@@ -1,21 +1,16 @@
...
-  <Launch>(
-    [
-      ...selectLaunches(Launches.UPCOMING),
-      ...selectLaunches(Launches.PREVIOUS),
-      ...selectLaunches(Launches.CREWED),
-      ...selectLaunches(Launches.CALIFORNIA),
-      ...selectLaunches(Launches.FLORIDA),
-    ].find((slide: Launch) => slide.id === selectActiveSlideId())
+export const selectLaunchDetails = (): Launch => {
+  const { launches }: { launches: Launch[] } = selectLaunches();
```
Comment (kerrimov) [inline src/components/LaunchDetailsPage/selectors/selectors.ts:12]: Providing type like this is not good idea

File: src/components/Home/reducers/launchReducer.ts [hunk 3]
```diff
@@ -1,35 +1,54 @@
...
-  [Launches.PREVIOUS]: [],
-  [Launches.CREWED]: [],
-  [Launches.CALIFORNIA]: [],
-  [Launches.FLORIDA]: [],
+const initState: LaunchState = {
+  launches: [],
+  launchAmount: 0,
   isLoading: false,
+  isScrollLoading: false,
   error: '',
```
Comment (kerrimov) [inline src/components/Home/reducers/launchReducer.ts:17]: A common alert will be used, so the error does not need here

File: src/components/Home/reducers/launchReducer.ts [hunk 4]
```diff
@@ -1,35 +1,54 @@
...
-        [filterValue]: payload[filterValue],
+        launches: checkPage(launches),
+        launchAmount,
         isLoading: false,
       };
     }
 
-    case `${filterValue}${LaunchActionTypes.LAUNCH_ERROR}`: {
+    case `${filterValue}_${LaunchActionTypes.LAUNCH_ERROR}`: {
+      const { error } = action.payload as LaunchActionError['payload'];
```
Comment (kerrimov) [inline src/components/Home/reducers/launchReducer.ts:51]: the same about types

File: src/components/Home/services/fetchLaunchConfig.ts [hunk 5]
```diff
@@ -1,9 +1,16 @@
...
 
-export const fetchLaunchConfig = (filterValue: string): FetchLaunchConfig => {
-  const params = { limit: 12 };
+export const fetchLaunchConfig = (
+  filterValue: string,
+  currentPage: number,
+): FetchLaunchConfig => {
+  const params: FetchLaunchConfig['params'] = {
+    limit: fetchDefaults.LIMIT,
+    offset: (currentPage - 1) * fetchDefaults.LIMIT,
```
[the remark this answers is not in the document]
Comment (vovkis13) [reply src/components/Home/services/fetchLaunchConfig.ts:12]: Using page counters more convinient then offset counters: all functions operate with pages, and just `fetchLaunchConfig` count offset

File: src/components/Home/reducers/launchReducer.ts (hunk 4 above)
Comment (vovkis13) [reply src/components/Home/reducers/launchReducer.ts]: line 51 deleted

File: src/components/Home/reducers/launchReducer.ts (hunk 3 above)
Comment (vovkis13) [reply src/components/Home/reducers/launchReducer.ts]: deleted

File: src/components/Home/actions/getLaunches.ts (hunk 1 above)
Comment (vovkis13) [reply src/components/Home/actions/getLaunches.ts]: LaunchResponse['data'] changed to new type `LaunchResponseData`

File: src/components/LaunchDetailsPage/selectors/selectors.ts (hunk 2 above)
Comment (vovkis13) [reply src/components/LaunchDetailsPage/selectors/selectors.ts]: I avoided destructuring and used `LaunchState` type

File: src/components/Home/components/LaunchList/LaunchList.tsx [hunk 6]
```diff
@@ -1,34 +1,73 @@
...
+  const dispatch: Dispatch<LaunchAction | ErrorAlertToggle> = useDispatch();
+  const launchListRef: RefObject<HTMLDivElement> = useOutletContext();
+
+  const { loadMoreRef, nextPage, setNextPage }: InfiniteScrollValues =
+    useInfiniteScroll();
+
+  const { launches, launchAmount, isLoading, isScrollLoading }: LaunchState =
+    selectLaunches();
+
+  const launchListIsEmpty: boolean = launches.length === 0;
```
[the remark this answers is not in the document]
Comment (kerrimov) [reply src/components/Home/components/LaunchList/LaunchList.tsx:33]: this comment a recommendation

**Review by kerrimov — APPROVED** [review]

**Review by kerrimov — APPROVED** [review]

**Pull request merged**